pkg/istio-agent/sds-agent.go

```diff
@@ -64,6 +64,7 @@ var (
 
 	pluginNamesEnv             = env.RegisterStringVar(pluginNames, "", "").Get()
 	enableIngressGatewaySDSEnv = env.RegisterBoolVar(enableIngressGatewaySDS, false, "").Get()
```

Comment:
Why do we need both ingress and egress? A single istio-agent is either ingress or egress? We should aim to keep config as minimal as possible

Reply:
Actually... now that I think about it it *can* be both ingress and egress. As a result we should actually likely have everything be a single config - same uds, etc. So we don't need to duplicate everything

Reply:
I think Security team prefers different configs(uds path etc.) for now, but I'll open an issue and follow up on that issue to merge both in a future PR. cc: @JimmyCYJ

Reply:
Why would we duplicate all of the code? its literally identical.

Reply:
Using a single uds sounds good to me. I would prefer to be conservative to begin with a different uds, to make sure we don't break the ingress sds logic in sds-agent. Later we can have a refactoring PR and combine two uds into one.

Reply:
So currently, the way we know if it's a ingress gateway proxy is by the existence of /var/run/ingress_gateway directory on the workload. Similarly for egress gw proxy... /var/run/egress_gateway. And most duplication ties back to the sdsudspaths, which I think to avoid any chance of regression we should keep separate for now and defer merging the two in a future PR. At least that's what I understood from the discussion I had with security team.

Reply:
Its not just "refactoring" - its a matter of correctness. There is, in reality, no such thing as "egress" and "ingress" gateways. There are just gateways. What if I want to configure a DR with TLS on my "ingress" gateway? I cannot do it because we decided we wanted to duplicate a bunch of code that is identical?

I also don't see why it would break the ingress logic - it is unchanged entirely. We just activate the codepath by default on egress gw now. Ingress remains unchanged

Reply:
Ingress and Egress gateway deployment are still two different resources even though they both tie back to the same Gateway API. I don't quite understand what you mean by DR with TLS on ingress gateway. Can you explain what case you're concerned about

Reply:
Lets not confuse the deployments we offer users as part of the default installation with the behavior of the system/APIs. A user can install arbitrary amounts of gateways, which are controlled by the Gateway API (among others). There is absolutely no distinction between "ingress" and "egress" gateways in Pilot code, because there is none in the API or in reality. 

There is client TLS (DR) and server TLS (Gateway) - there is not "egressgateway" tls and "ingressgateway" tls. We can either, or both, on any gateway.

Just because we have a default example of a split ingress and egress gateway does not mean that is how gateways actually work. They are merely examples/best practices to get people started.

Reply:
If I understand your case correctly, I think the case you are mentioning is ingress gw originating tls, well if that's the case- this feature isn't meant to support that use case currently.
external workload -> ingress gateway -> external workload is not a use case I think we ever see? 
For external workload -> ingress gateway -> sidecar, we aren't even supporting DR to use credentialName to fetch secrets at ingress gateway.

Reply:
> this feature isn't meant to support that use case currently.

Then I disagree with this feature :slightly_smiling_face: . We cannot just merge stuff that applies to small subset of cases - its a horrible user experience to try to debug your way through all of the partially implemented APIs.

You're examples are too limited - applying strict assumptions to a project with as wide a user base as Istio will not hold up in practice. Its not `external workload -> ingress gateway -> external workload`. The flow is `arbitrary workload -> arbitrary gateway -> arbitrary workload`

Reply:
Ah I agree with you now, my examples use case was too restrictive. `arbitrary workload -> arbitrary gateway -> arbitrary workload` would work best if we use single uds path.

Reply:
That's good point. I think what we need is to rename the uds path to be a general one, and let sds-server listens on this uds path, and update the logic in ingressSdsExists() when creating gatewaySecretCache.

Reply:
The actual name is less important imo. I don't think a user will complain if some internal name is called `ingress_uds` and its on their egress gateway. Changing the name has real costs (its not currently dynamically configurable) so probably not worth it.